#11616 disttrial worker test improvements

exarkun opened the pull request:
Fixes #11616

exarkun committed:
add HasSum, IsSequenceOf, isFailure, similarFrame, and isTuple matchers

These are handy for writing hamcrest-based tests for a lot of trial code.

exarkun committed:
Fix WorkerProtocolTests and LocalWorkerAMPTests

The previous implementation used unnecessary fakes and encoded unnecessary
implementation details as well as unnecessarily duplicating in-memory
ITransport implementation available in other testing libraries in Twisted.

This is an almost-no-behavior-change commit that just makes the tests simpler
and exercises the code under test in a more realistic way.

exarkun committed:
Tests for isTuple

exarkun committed:
Tests for isFailure and some bug fixes

exarkun committed:
news fragment

adiroiban commented on src/twisted/trial/_dist/test/test_worker.py:
```diff
@@ -86,197 +76,155 @@ class LocalWorkerAMPTests(TestCase):
         class StopTestResult(TestResult):
-            def stopTest(self, test):
-                stopped.append(test)
+            _stopped = False
+
+            def stopTest(self, test: PyUnitTestCase) -> None:
+                self._stopped = True
```
```suggestion
            _stopped = []

            def stopTest(self, test: PyUnitTestCase) -> None:
                self._stopped.append(True)
```

I am still using append in tests at it can help detect corner cases in which something is called multiple times.

The append might look strange on a class variable, but this class should only be used once, inside this test.

The test code looks uglier ... but I think is best to have as strict of an assertion as possible.

The test code should be easy to maintain, but at the same time is there to prevent unexpected behaviour.

exarkun replied:
Perhaps an integer counter?

exarkun replied:
I put the list back because the original test also asserted the right test case was passed as an argument.

adiroiban commented on src/twisted/trial/_dist/test/test_worker.py:
```diff
@@ -86,197 +76,155 @@ class LocalWorkerAMPTests(TestCase):
+                        ),
+                    ),
+                ),
+            ),
```
I am not a big fan of these generic assertions.

we can leave this as it is, since `matched_result` is already in trunk.
I find the name a bit too generic.

Something to my taste.
```
assert_than(result, is_testrun_failure(case, type=AssertionError, value=WorkerException("pyunit failure")
```

and all the extra `isTuple`, `equal_to` and the general structure is handled by the dedicated matcher.

From my experience, with these specialiez markers it was easier to refactor the internal structure of a code as you might only need to update the specialized marker, without updating each test.

exarkun replied:
This overlaps a bit with the comment above and I left a reply there.

adiroiban commented on src/twisted/trial/_dist/test/test_worker.py:
```diff
@@ -86,197 +76,155 @@ class LocalWorkerAMPTests(TestCase):
-            frames=["file.py", "invalid code", "3"],
+        case = pyunitcases.PyUnitTest("test_error")
+        result = self.workerRunTest(case)
+        assert_that(
```
at least for me, this embedded structure is not easier to read ... 
but maybe is just my bias for old-school xunit type assertions.

```diff
- self.assertEqual(failure.type, ValueError)
- self.assertEqual(failure.value, WorkerException("error"))
+ self.assertTestRunFailure(self.result, ValueError, WorkerException("error"))
```

For this type of assertion, you only focus on the end/high-level result, and ignore the internal structure of a test case.

It might be an advantage or disadvantage.

I usually write a test in which the internal structure is validated, and then all the other tests only check the high-level values, ignore the internal structure.


With the generic `matches_result` you need to remember the full structure of the test case result, each time you write the test.

exarkun replied:
> at least for me, this embedded structure is not easier to read ...

I don't object to that.  I think a lot of it is about familiarity and experience but it is also true that matcher-style tends to have fewer larger expressions rather than more smaller expressions.

Larger expressions in this style probably often take more practice to become comfortable reading.  The benefit that you get is when multiple parts of the value are wrong, you get to learn about all of them after one test run instead of only the one related to the very first assertion.

testtools allows the best of both worlds with an `expectThat` function which can add a test failure _and_ let the test keep running.  But Twisted decided to go with pyhamcrest instead...

Anyhow, I'm open to improving the readability of these tests before merging.

exarkun replied:
> + self.assertTestRunFailure(self.result, ValueError, WorkerException("error"))

Looking at this idea specifically in more detail, I don't see a good way to make it work.  The trouble is that `assertTestRunFailure` isn't reusable.  `test_runError`, `test_runFailure`, `test_runSkip`, etc all look similar but they differ in slight ways.

If we rename the suggested function to `assertTestRunError` (since this case is about test suite "errors" rather than "failures") and we implemented it _however_ (doesn't matter if we use hamcrest or not) then it is usable by exactly one test, `test_runError`.  So we need another one for `test_runFailure`, and another for `test_runSkip`, etc.

In this case I don't think there's any benefit to abstracting the details away from the test.  The purpose of the test method is exactly to encapsulate these details.

I can flatten the hamcrest code so it does repeated calls to `assert_that` instead of a single call with a complex matcher.  I think that change makes it easier for folks without a lot of practice writing and reading these and it doesn't hurt runtime properties _that_ much probably.

I don't think it makes sense to take the details of those assertions out of the test methods though.  Or at least, I don't see a way that doesn't just lead to more work at development time and more work for future readers of the code.

exarkun replied:
I flattened the matcher in https://github.com/twisted/twisted/pull/11617/commits/6deb9f1bd6a237f3792ecc7e39f5e6d01f457fba - I'm curious if you find that to be any better

adiroiban replied:
thanks for the info and the updated.
Agree.
The flattened version looks easier to read, at a first read.

I am sure that after reading and looking at the code a few more times, the nested structures will look much easier to read.
I just wanted to leave a "raw" feedback.

I guess that other people might want to help contribute the Twisted, but then they found some code that they don't understand and then they give up.

I am searching for the formula for creating an open source project that is easy to contribute to, while still keeping high code quality standards.

adiroiban reviewed (approved):
Thanks for the refactoring.

This is not my style of writing tests... and I am not happy with `matches_result`..and this method is already in trunk.

Overall, there are a few nice cleanups for the tests and I think is best to have it merged.

adiroiban reviewed (changes requested):
I now see that this PR was not ready for review.

--------

On a general note, I am -1 for using hypothesis in twisted, especially for general code.
It adds another thing that you must learn and be familiar with, before contributing to the project.

And for people not very familiar  with the library, might be a reason to write invalid tests.

exarkun committed:
use typing.List to support more Python versions

exarkun committed:
switch to typing_extensions.Protocol to support more Python versions

exarkun committed:
Merge remote-tracking branch 'origin/trunk' into 11616-test_worker-improvements

exarkun committed:
some changes which hopefully improve readability

exarkun committed:
Side-step pyhamcrest's nonsensical types

exarkun committed:
restore more of the original intent of this test

make sure stopTest is called only once and with the right argument

exarkun committed:
flatten the matcher in test_runError to see if it helps readability

exarkun commented:
needs-review

adiroiban commented on src/twisted/trial/_dist/worker.py:
```diff
@@ -185,7 +185,7 @@ def addExpectedFailure(
         Add an expected failure to the reporter.
         """
-        _todo = Todo(todo)
+        _todo = Todo("<unknown>" if todo is None else todo)
```
I guess that this change was done to make it easier to remove `Todo.__init__` while defining `reason: str` and not `Optional[str]`

I am happy to always have a string as the reason.

exarkun replied:
Yep.  I also took a hint from the existing docstring for `Todo` which claimed `reason` was a string and said nothing about None.

adiroiban reviewed (approved):
I think that this can be merged.

------

I am not a big fan of hamcrest , but I am not  -1 on it.

Jean-Paul, if you think that testtool.assertThat provides a better dev experience in terms of writing matches and diffs, I think we can still consider a switch.

---------

Using simple hypothesis strategies can make the code easier to use and encourage developers to write tests with explicit types.

I saw many bugs (in python 2.7) that were not discovered by a test, because the test was using something like `received = 'binary-data'` ... but that `binary-data` was a valid text that survived an accidental UTF-8 decode on it. 
And in prod you have real binary data and then UTF-8 decoding fails.

-----------

I only did a quick review.

I don't know if anyone will have time to review it.
I think is best to have it merge, than keep this PR in the review queue for many weeks

exarkun committed:
Flatten the nested matches into several simpler matches

This also removes the need for isTuple

exarkun committed:
remove unused imports

exarkun committed:
simplify one more hamcrest usage

exarkun committed:
type annotate `makeTodo`

exarkun committed:
Force the correct types for trial's TestResult

They disagree with the inherited type which makes TestResult a very poorly
behaved subclass.  But, so it is.

exarkun commented:
> thanks for the info and the updated.
> Agree.
> The flattened version looks easier to read, at a first read.
> 
> I am sure that after reading and looking at the code a few more times, the nested structures will look much easier to read.
> I just wanted to leave a "raw" feedback.
> 
> I guess that other people might want to help contribute the Twisted, but then they found some code that they don't understand and then they give up.
> 
> I am searching for the formula for creating an open source project that is easy to contribute to, while still keeping high code quality standards.
> 

Thanks for the reviews @adiroiban .  I went ahead and flattened the hamcrest usage in the other `test_worker.py` tests to match the example I pushed earlier.  I don't mind trying to keep these expressions more on the simple side.  I think you make a good point about the consequences of such complex expressions on new contributors (or even existing contributors not familiar with this testing style).

I think having a richer matching library can also help us out and right now we're just starting to build one up for trial so we have some more work to do in that department.  Also, I think "it's hard to describe what result I want" is a bit of a code smell for "what you want is not great" so as we find it is hard to write good matchers we might decide we want to introduce better data structures into the implementation to make it easier.

exarkun committed:
Old Python compat

exarkun committed:
Merge remote-tracking branch 'origin/trunk' into 11616-test_worker-improvements

exarkun committed:
Turn on remote debugging

exarkun committed:
do tmate first

exarkun committed:
some debug junk

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

exarkun committed:
don't twiddle stdout, it makes debugging hard

the broken test doesn't try to write to sys.stdout so it doesn't matter (I hope!!)

exarkun committed:
Merge remote-tracking branch 'mine/11616-test_worker-improvements' into 11616-test_worker-improvements

exarkun committed:
breaks other tests, hampers reproduction

exarkun committed:
better failure messages

exarkun committed:
let tests run first so tox initializes the environment, but end the test run early

exarkun committed:
dump more debug info

exarkun committed:
not really usable

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

exarkun committed:
maybe eek out a bit more info

exarkun committed:
Merge remote-tracking branch 'mine/11616-test_worker-improvements' into 11616-test_worker-improvements

exarkun committed:
debug stuff interferes with test logic!  rad

exarkun committed:
remove debug stuff

the failure went away on CI.  will it still be gone in the next run?  who can say

exarkun committed:
Merge remote-tracking branch 'origin/trunk' into 11616-test_worker-improvements

exarkun committed:
Merge branch '11638.manhole-vs-excepthook' into 11616-test_worker-improvements

exarkun committed:
Merge branch 'trunk' into 11616-test_worker-improvements
